keras_nlp/models/bert.py

```diff
@@ -0,0 +1,287 @@
+            The hidden size must be divisible by the number of attention heads.
+        intermediate_dim: The output dimension of the first Dense layer in a
+            two-layer feedforward network for each transformer.
+        dropout: Dropout probability for the Transformer encoder.
+        max_sequence_length: The maximum sequence length that this encoder can
+            consume. If None, max_sequence_length uses the value from sequence
+            length. This determines the variable shape for positional
+            embeddings.
+        num_segments: The number of types that the 'segment_ids' input can
+            take.
+
+    Example usage:
+    ```python
+    # Randomly initialized Bert encoder
+    encoder = keras_nlp.models.Bert(
+        vocab_size=30522,
+        num_layers=12,
+        hidden_size=768,
+        num_heads=12,
+        intermediate_dim=3072,
+        dropout=0.1,
+        max_sequence_length=12
+    )
+
+     # Call encoder on the inputs.
+    input_data = {
+        "input_ids": tf.random.uniform(
+            shape=(1, 12), dtype=tf.int64, maxval=30522),
+        "segment_ids": tf.constant(
+            [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)),
+        "input_mask": tf.constant(
+            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)),
+    }
+    output = encoder(input_data)
+    ```
+    """
+
+    # TODO(bischof): add tests
+
+    def __init__(
+        self,
+        vocab_size,
+        num_layers,
+        hidden_size,
+        num_heads,
+        intermediate_dim,
+        dropout=0.1,
+        max_sequence_length=512,
+        num_segments=2,
+        **kwargs,
+    ):
+
+        initializer_fn = keras.initializers.TruncatedNormal(stddev=0.02)
+
+        # Functional version of model
+        token_id_input = keras.Input(
+            shape=(None,), dtype="int32", name="input_ids"
+        )
+        segment_id_input = keras.Input(
+            shape=(None,), dtype="int32", name="segment_ids"
+        )
+        input_mask = keras.Input(
+            shape=(None,), dtype="int32", name="input_mask"
+        )
+
+        # Embed tokens, positions, and segment ids.
+        token_embedding_layer = keras.layers.Embedding(
+            input_dim=vocab_size,
+            output_dim=hidden_size,
+            name="token_embedding",
+        )
+        token_embedding = token_embedding_layer(token_id_input)
+        position_embedding = PositionEmbedding(
+            initializer=initializer_fn,
+            sequence_length=max_sequence_length,
+            name="position_embedding",
+        )(token_embedding)
+        segment_embedding = keras.layers.Embedding(
+            input_dim=num_segments,
+            output_dim=hidden_size,
+            name="segment_embedding",
+        )(segment_id_input)
+
+        # Sum, normailze and apply dropout to embeddings.
+        x = keras.layers.Add(
+            name="embeddings/sum",
```

Comment:
Naming comment: `/` is normally used to denote layer hierarchy. Here the `Add`/`LayerNormalization` are not sublayers of `Embedding` so maybe better to go with `embeddings_sum`, etc.

Reply:
Also note that stateless layers like `Add` typically don't need explicit names. The autogenerated name is fine.

Reply:
Removed explicit name for `Add` and replaced slashes with underscores elsewhere